arangod/Pregel/Messaging/Aggregate.h

```diff
@@ -0,0 +1,75 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2014-2022 ArangoDB GmbH, Cologne, Germany
+/// Copyright 2004-2014 triAGENS GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Julia Volmer
+////////////////////////////////////////////////////////////////////////////////
+#pragma once
+
+#include <cstdint>
+#include <optional>
+#include <utility>
+#include "Basics/ResultT.h"
+
+namespace arangodb::pregel {
+
+template<typename T>
+concept Addable = requires(T a, T b) {
+  {a.add(b)};
+};
+
+template<Addable T>
+struct Aggregate {
+  Aggregate(T initialValue, uint64_t countUntilFinished)
+      : _countUntilFinished{countUntilFinished},
+        _aggregate{std::move(initialValue)} {}
+  static auto withComponentsCount(uint64_t countUntilFinished) -> Aggregate {
+    return Aggregate{T{}, countUntilFinished};
+  }
+  auto aggregate(T const& message) -> std::optional<T> {
```

Comment:
This will copy `message` in the next line if `_aggregate.add(message)` needs to save the message. `auto aggregate(T message)` would allow the caller to pass it as an r-value and to move it. Also `_aggregate.add()` could move it then. Another option is to define both the above and `aggregate(T&& message)` with the same behavior. (@markuspf is it correct?)

Reply:
Here, I think `aggregate(T message)` is the correct thing to use.

using `aggregate(T&& message)` forces the caller to use `std::move` for lvalues; this can be used to make it more obvious that an input is indeed moved.

(I find this here to be a fairly good reference-entrypoint: https://www.foonathan.net/2018/03/rvalue-references-api-guidelines/)